### PR TITLE
improved intersection performance

### PR DESCRIPTION
**Pull request by landvibe**
Intersection can be improved by sorted array.
The time complexity
- previous code: `m*n^2`
- improved code: `m*n*logn`

below is my test result on my machine
- two arrays with 100,000 items
- previous code: 16830 ms
- improved code: 93 ms

### CODE REVIEW

**Comment by megawac** [comment]
A solution more likely to be accepted would be to implement a side case using [ES2015 Sets](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Set)

```js
let set = new Set(array);

// ...

set = new Set(_.filter(inputList, x => set.has(x)));
```

**Comment by landvibe** [comment]
@megawac 
MDN says "Set objects are collections of values, you can iterate its elements in insertion order"
I think it is not sorted data structure.

**Comment by megawac** [comment]
Yes, `intersection` is supposed to preserve original order. The PR as is is a breaking change.

```js
_.intersection([2, 3,1], [1,2])

```

File: underscore.js [hunk 1]
```diff
@@ -571,16 +571,32 @@
...
+        if (_.indexOf(sortedList, inputList[j], true) >= 0) {
+          tempList.push(inputList[j]);
+        }
+      }
+      sortedList = tempList;
+    }
+    // preserves the order of the first array
+    // and get a duplicate-free array
+    var occurMap = [];
+    for (i = 0, length = array.length; i < length; i++) {
```
Comment (landvibe) [inline underscore.js]: @megawac 
The PR preserves original order

**Comment by landvibe** [comment]
I fixed a bug
bug: wrong output when there are more than 2 arrays
and I added new test case

**Comment by jgonggrijp** [comment]
It's a nice idea but it only works for sortable values (i.e., strings and numbers). It has to work with unsortable values, too.

**Pull request closed without merging**